server/src/test/java/org/elasticsearch/index/mapper/RuntimeAliasTests.java

```diff
@@ -0,0 +1,106 @@
+        MapperService mapperService = createMapperService(topMapping(b -> {
+            b.startObject("runtime");
+            {
+                b.startObject("alias-to-field").field("type", "alias").field("path", "field").endObject();
+            }
+            b.endObject();
+        }));
+        Exception e = expectThrows(IllegalStateException.class, () -> mapperService.mappingLookup().getFieldType("alias-to-field"));
+        assertEquals("Cannot resolve alias [alias-to-field]: path [field] does not exist", e.getMessage());
+    }
+
+    public void testAliasToAlias() throws IOException {
+        MapperService mapperService = createMapperService(topMapping(b -> {
+            b.startObject("runtime");
+            {
+                b.startObject("alias-to-field").field("type", "alias").field("path", "field").endObject();
+                b.startObject("alias-to-alias").field("type", "alias").field("path", "alias-to-field").endObject();
+            }
+            b.endObject();
+            b.startObject("properties");
+            {
+                b.startObject("field").field("type", "keyword").endObject();
+            }
+            b.endObject();
+        }));
+
+        MappedFieldType aliased = mapperService.mappingLookup().getFieldType("alias-to-alias");
+        assertEquals("field", aliased.name());
+        assertEquals(KeywordFieldMapper.KeywordFieldType.class, aliased.getClass());
+    }
+
+    public void testAliasLoops() throws IOException {
+        MapperService mapperService = createMapperService(topMapping(b -> {
+            b.startObject("runtime");
+            {
+                b.startObject("alias-loop1").field("type", "alias").field("path", "alias-loop2").endObject();
+                b.startObject("alias-loop2").field("type", "alias").field("path", "alias-loop1").endObject();
+            }
+            b.endObject();
+            b.startObject("properties");
+            {
+                b.startObject("field").field("type", "keyword").endObject();
+            }
+            b.endObject();
+        }));
+
+        Exception e = expectThrows(IllegalStateException.class, () -> mapperService.mappingLookup().getFieldType("alias-loop1"));
+        assertEquals("Loop in field resolution detected: alias-loop1->alias-loop2->alias-loop1", e.getMessage());
+    }
+
+    public void testDynamicLookup() throws IOException {
+        MapperService mapperService = createMapperService(topMapping(b -> {
+            b.startObject("runtime");
+            {
+                b.startObject("dynamic-alias").field("type", "alias").field("path", "flattened").endObject();
+            }
+            b.endObject();
+            b.startObject("properties");
+            {
+                b.startObject("flattened").field("type", "flattened").endObject();
+            }
+            b.endObject();
+        }));
+
+        MappedFieldType dynamic = mapperService.fieldType("flattened.key");
```

Comment:
Small comment, we could check the subfield paths too (`KeyedFlattenedFieldType#key`)?